src/graphql/utils.ts

```diff
@@ -1,24 +1,132 @@
+}
 
-const DEFAULT_CONTENT_TYPE_FIELDS = ['sys', '__typename', 'contentfulMetadata'];
+function getTypeNameForNode(
+  node: FragmentDefinitionNode | OperationDefinitionNode
+): string | undefined {
+  return 'typeCondition' in node ? node.typeCondition.name.value : node.name?.value;
+}
+
+interface GeneratedGraphQLStructure {
+  name: string;
+  alias?: string;
+  __typename: string;
+}
 
-export function logUnrecognizedFields(contentTypeFields: string[], data: Entity): void {
-  const recognized = new Set([...DEFAULT_CONTENT_TYPE_FIELDS, ...contentTypeFields]);
+function detectFieldAlias(
+  selections: Readonly<SelectionNode[]>,
+  typename: string
+): GeneratedGraphQLStructure[] {
+  const generated: GeneratedGraphQLStructure[] = [];
 
-  for (const field of Object.keys(data)) {
-    if (!(recognized.has(field) || recognized.has(removeCollection(field)))) {
-      debug.warn(`Unrecognized field '${field}'. Note that GraphQL aliases are not supported`);
+  for (const selection of selections) {
+    if (selection.kind === 'Field') {
+      generated.push({
+        name: selection.name.value,
+        alias: selection.alias?.value,
+        __typename: typename,
+      });
+
+      if (selection.selectionSet?.selections) {
+        generated.push(
+          ...detectFieldAlias(selection.selectionSet.selections, selection.name.value)
+        );
+      }
     }
   }
+
+  return generated;
 }
 
-const COLLECTION_SUFFIX = 'Collection';
+export function parseGraphQLParams(query: DocumentNode): GraphQLParams {
+  const generated: GeneratedGraphQLStructure[] = [];
 
-export function buildCollectionName(name: string): string {
-  return `${name}${COLLECTION_SUFFIX}`;
+  for (const def of query.definitions) {
+    if (def.kind === 'OperationDefinition' || def.kind === 'FragmentDefinition') {
+      const typename = getTypeNameForNode(def);
+
+      if (!typename) {
+        debug.warn('Could not generate __typename for query definition', def);
+        continue;
+      }
+
+      for (const selection of def.selectionSet.selections) {
+        if (selection.kind === 'Field') {
+          generated.push(...detectFieldAlias(def.selectionSet.selections, typename));
+        }
+      }
+    }
+  }
+
+  const params: GraphQLParams = new Map();
+  for (const { __typename, alias, name } of generated) {
+    const match = params.get(__typename) || {
+      alias: new Map<string, string>(),
+      fields: new Set<string>(),
+    };
+
+    // Add the field name
+    match.fields.add(name);
+    // Add the alias
+    if (alias) {
+      match.alias.set(name, alias);
+    }
+
+    params.set(__typename, match);
+  }
+
+  return params;
+}
+
+function getAliasForFieldName(name: string, typename: string, gqlParams?: GraphQLParams): string {
+  return gqlParams?.get(typename)?.alias.get(name) || name;
 }
 
-function removeCollection(name: string): string {
-  return name.replace(COLLECTION_SUFFIX, '');
+export function isRequestedField(
```

Comment:
Without the context I feel like this function could be hard to understand in future. Possibly something like
`isFieldInQuery` could be more explicit here